assembly/src/parser/mod.rs

```diff
@@ -159,6 +159,176 @@ fn parse_forms_internal(
         .map_err(|err| ParsingError::from_parse_error(source_id, err))
 }
 
+// DIRECTORY PARSER
+// ================================================================================================
+
+/// Read the contents (modules) of this library from `dir`, returning any errors that occur
+/// while traversing the file system.
+///
+/// Errors may also be returned if traversal discovers issues with the modules, such as
+/// invalid names, etc.
+///
+/// Returns an iterator over all parsed modules.
+#[cfg(feature = "std")]
+pub fn read_modules_from_dir(
+    namespace: crate::LibraryNamespace,
+    dir: &std::path::Path,
+    source_manager: Arc<dyn SourceManager>,
+) -> Result<impl Iterator<Item = Box<ast::Module>>, Report> {
+    use std::collections::{btree_map::Entry, BTreeMap};
+
+    use module_walker::{ModuleEntry, WalkModules};
+
+    use crate::diagnostics::{IntoDiagnostic, WrapErr};
+
+    if !dir.is_dir() {
+        return Err(Report::msg(format!(
+            "the provided path '{}' is not a valid directory",
+            dir.display()
+        )));
```

Comment:
```suggestion
        return Err(miette!(
            "the provided path '{}' is not a valid directory",
            dir.display()
        ));
```

This can be applied elsewhere in the file too where we also use `Report::msg(format!(...))`

Reply:
Good idea! Replaced here and in other files in the assembler.